src/transformStateWithClones.js

```diff
@@ -7,7 +7,43 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const listOfStates = [];
+  const newState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        addProperties(newState, action.extraData);
+        listOfStates.push({ ...newState });
+        break;
+      case 'removeProperties':
+        removeProperties(newState, action.keysToRemove);
+        listOfStates.push({ ...newState });
+        break;
+      case 'clear':
+        clear(newState);
+        listOfStates.push({});
+        break;
+    }
+  }
+
+  return listOfStates;
+}
+
+function addProperties(state, extraData) {
+  Object.assign(state, extraData);
+}
+
+function removeProperties(state, keysToRemove) {
+  for (const key of keysToRemove) {
+    delete state[key];
+  }
```

Comment:
The 'removeProperties' function directly mutates the passed 'state' object. To avoid side effects, create a copy of the state before removing properties.